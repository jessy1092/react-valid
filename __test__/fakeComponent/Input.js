import React from 'react';

const Input = ({ validate, message }) => (
  <div>
    <input onChange={e => validate(e.target.value)} />
    <div>{message}</div>
  </div>
);

export default Input;
