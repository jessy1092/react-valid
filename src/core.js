import React from 'react';

export const addMethod = (validators, name, method = () => {}, message = '') => ({
  ...validators,
  [name]: {
    method,
    message,
  },
});

export const addMethods = (validators, appendValidators = {}) => {
  const defaultValidator = {
    method() {},
    message: '',
  };

  const newValidates = {
    ...validators,
  };

  Object.keys(appendValidators).forEach(name => {
    newValidates[name] = {
      ...defaultValidator, // Set default value to prevent method not define
      ...appendValidators[name],
    };
  });

  return newValidates;
};

export const connect = (validators, Component) => {
  const displayName = Component.displayName || Component.name || 'Component';

  class ValidComponent extends React.Component {
    constructor(props) {
      super(props);
      this.state = {
        valid: true,
        message: '',
      };

      this.validate = this.validate.bind(this);
    }

    validate(value) {
      const { onError = () => {}, ...other } = this.props;
      let message = '';

      const status = Object.keys(validators).every(name => {
        if (Object.prototype.hasOwnProperty.call(other, name)) {
          if (!validators[name].method(value, this.props[name])) {
            if (typeof validators[name].message === 'string') {
              // Display normal string message
              message = validators[name].message;
            } else {
              // Use template to generate message
              message = validators[name].message(value, this.props[name]);
            }
            return false;
          }
        }
        return true;
      });

      this.setState(
        {
          valid: status,
          message,
        },
        () => {
          if (!status) {
            onError(value);
          }
        },
      );
    }

    render() {
      return <Component {...this.props} {...this.state} validate={this.validate} />;
    }
  }
  ValidComponent.displayName = `Valle(${displayName})`;

  return ValidComponent;
};

export const template = (strings, ...keys) => (validateValue, props) => {
  const result = [strings[0]];
  keys.forEach((key, i) => {
    let value = '';
    if (key === 0) {
      value = validateValue;
    } else if (key === 1) {
      value = typeof props === 'string' || !isNaN(props) ? props : props[key];
    } else {
      value = props[key];
    }
    result.push(value, strings[i + 1]);
  });
  return result.join('');
};
