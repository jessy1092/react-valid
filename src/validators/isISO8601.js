import isISO8601 from 'validator/lib/isISO8601';

import { template } from '../core';

import { isEmpty } from './util';

export default {
  isISO8601: {
    method: value => isEmpty(value) || isISO8601(value),
    message: template`${0} is not a valid ISO 8601 date. Please check your value and try again.`,
  },
};
