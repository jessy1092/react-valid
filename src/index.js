import { addMethod, addMethods, connect, template } from './core';

import isEmail from './validators/isEmail';
import contains from './validators/contains';
import equals from './validators/equals';
import isBoolean from './validators/isBoolean';
import isFloat from './validators/isFloat';
import isISO8601 from './validators/isISO8601';
import isInt from './validators/isInt';

export function createValle() {
  let validators = {};

  return {
    addMethod(name, method, message) {
      validators = addMethod(validators, name, method, message);
    },
    addMethods(appendValidators) {
      validators = addMethods(validators, appendValidators);
    },
    connect(Component) {
      return connect(validators, Component);
    },
    template,
  };
}

const valle = createValle();

valle.addMethods(isEmail);
valle.addMethods(contains);
valle.addMethods(equals);
valle.addMethods(isBoolean);
valle.addMethods(isFloat);
valle.addMethods(isISO8601);
valle.addMethods(isInt);

export default valle;
