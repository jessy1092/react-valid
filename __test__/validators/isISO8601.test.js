import React from 'react';
import { mount } from 'enzyme';

import { createValle } from '../../src';
import isISO8601 from '../../src/validators/isISO8601';

import Input from '../fakeComponent/Input';

test('It sould be invalid if input is not ISO8601 date', () => {
  const valle = createValle();
  valle.addMethods(isISO8601);

  const ValidInput = valle.connect(Input);
  const wrapper = mount(<ValidInput isISO8601 />);

  wrapper.find('input').simulate('change', {
    target: {
      value: 'Hi',
    },
  });

  expect(wrapper.state('valid')).toEqual(false);
  expect(wrapper.state('message')).toEqual(
    'Hi is not a valid ISO 8601 date. Please check your value and try again.',
  );
  expect(wrapper).toMatchSnapshot();
});

test('It sould be valid if input is ISO8601 date', () => {
  const valle = createValle();
  valle.addMethods(isISO8601);

  const ValidInput = valle.connect(Input);
  const wrapper = mount(<ValidInput isISO8601 />);

  wrapper.find('input').simulate('change', {
    target: {
      value: '2017-04-26T12:30:48,3-06:00',
    },
  });

  expect(wrapper.state('valid')).toEqual(true);
  expect(wrapper.state('message')).toEqual('');
  expect(wrapper).toMatchSnapshot();
});
