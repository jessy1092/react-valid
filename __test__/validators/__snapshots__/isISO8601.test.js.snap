// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`It sould be invalid if input is not ISO8601 date 1`] = `
<div>
  <input />
  <div>
    Hi is not a valid ISO 8601 date. Please check your value and try again.
  </div>
</div>
`;

exports[`It sould be valid if input is ISO8601 date 1`] = `
<div>
  <input />
  <div />
</div>
`;

exports[`It sould be valid if input is empty 1`] = `
<div>
  <input />
  <div />
</div>
`;
